fdbserver/StorageCache.actor.cpp

```diff
@@ -57,46 +864,141 @@ ACTOR Future<Void> pullAsyncData( StorageCacheData *self ) {
+					case MutationRef::ByteMax:
+					case MutationRef::ByteMin:
+					case MutationRef::Max:
+					case MutationRef::Min:
+					case MutationRef::MinV2:
+					case MutationRef::Or:
+					case MutationRef::Xor:
+					case MutationRef::CompareAndClear:
+						++data->counters.atomicMutations;
+						break;
+				}
+			}
+			else
+				TraceEvent(SevError, "DiscardingPeekedData", data->thisServerID).detail("Mutation", msg.toString()).detail("Version", r->version().toString());
 
 			tagAt = r->version().version + 1;
 		}
 
+		if(ver != invalidVersion) {
+			data->lastVersionWithData = ver;
+		} else {
+			// TODO understand this better
+			ver = r->version().version - 1;
+		}
+
+		//if(ver == invalidVersion && !fii.changes.empty() ) {
+		//	ver = updater.currentVersion;
+		//}
+
+		if(ver != invalidVersion && ver > data->version.get()) {
+			debugKeyRange("SCUpdate", ver, allKeys);
+
+			data->mutableData().createNewVersion(ver);
+
+			// TODO what about otherError
+			if (data->otherError.getFuture().isReady()) data->otherError.getFuture().get();
+
+			// TODO may enable these later
+			//data->noRecentUpdates.set(false);
+			//data->lastUpdate = now();
+			data->version.set( ver );		// Triggers replies to waiting gets for new version(s)
+			// TODO double check
+			//setDataVersion(data->thisServerID, data->version.get());
+
+			// TODO what about otherError
+			if (data->otherError.getFuture().isReady()) data->otherError.getFuture().get();
+
+			// we can get rid of versions beyond maxVerionsInMemory at any point. Update the
+			//desiredOldestVersion and that may invoke the compaction actor
+			Version maxVersionsInMemory = SERVER_KNOBS->MAX_READ_TRANSACTION_LIFE_VERSIONS;
+			Version proposedOldestVersion = data->version.get() - maxVersionsInMemory;
+			proposedOldestVersion = std::max(proposedOldestVersion, data->oldestVersion.get());
+			data->desiredOldestVersion.set(proposedOldestVersion);
+		}
+
+		// TODO
+		//validate(data);
+
+		if(r->version().version >= data->lastTLogVersion) {
+			if(data->behind) {
+				TraceEvent("StorageCacheNoLongerBehind", data->thisServerID).detail("CursorVersion", r->version().version).detail("TLogVersion", data->lastTLogVersion);
+			}
+			data->behind = false;
+		}
+
 		tagAt = std::max( tagAt, r->version().version);
 	}
 }
 
 ACTOR Future<Void> storageCache(StorageServerInterface ssi, uint16_t id, Reference<AsyncVar<ServerDBInfo>> db) {
-	state StorageCacheData self(ssi.id());
-	state PromiseStream<Future<Void>> addActor;
-	state Future<Void> error = actorCollection( addActor.getFuture() );
+	state StorageCacheData self(ssi.id(), id);
+	//state PromiseStream<Future<Void>> addActor;
+	state ActorCollection actors(false);
+	//state Future<Void> error = actorCollection( addActor.getFuture() );
 	state Future<Void> dbInfoChange = Void();
 
-	addActor.send(pullAsyncData(&self));
+	self.ck = cacheKeysPrefixFor( id ).withPrefix(systemKeys.begin);  // FFFF/02cacheKeys/[this server]/
```

Comment:
Still needs to pull data for the ranges this cache is responsible for

Reply:
Not sure if I am missing something in my understanding. I moved the actor to pull the mutations to line 969. So we will pull all mutations with the cache tag as you coded. This key helps figure out the private mutations (for the cached key-range) that a cache server should process.